qemu/tests/migration_with_block.py

```diff
@@ -0,0 +1,233 @@
+            test.fail("Fail to change cdrom to %s." % new_img_name)
+
+    def change_vm_power():
+        """ Change the vm power. """
+        method, command = params['command_opts'].split(',')
+        logging.info('Sending command(%s): %s' % (method, command))
+        if method == 'shell':
+            p_session = vm.wait_for_login(timeout=360)
+            p_session.sendline(command)
+            p_session.close()
+        else:
+            getattr(vm.monitor, command)()
+
+    def check_vm_status(timeout=600):
+        """ Check the status of vm. """
+        action = 'shutdown' if shutdown_vm else 'login'
+        if not getattr(vm, 'wait_for_%s' % action)(timeout=timeout):
+            test.fail('Failed to %s vm.' % action)
+
+    def set_dst_params():
+        """ Set the params of dst vm. """
+        for name, val in ast.literal_eval(params.get('set_dst_params', '')).items():
+            vm.params[name] = val
+
+    shutdown_vm = params.get('shutdown_vm', 'no') == 'yes'
+    reboot = params.get('reboot_vm', 'no') == 'yes'
+    src_desc = params.get('src_addition_desc', '')
+    dst_desc = params.get('dst_addition_desc', '')
+    mig_timeout = float(params.get("mig_timeout", "3600"))
+    mig_protocol = params.get("migration_protocol", "tcp")
+    mig_cancel_delay = int(params.get("mig_cancel") == "yes") * 2
+    inner_funcs = ast.literal_eval(params.get("migrate_inner_funcs", "[]"))
+    capabilities = ast.literal_eval(params.get("migrate_capabilities", "{}"))
+    with_cdrom = params.get('with_cdrom', 'no') == 'yes'
+    os_type = params['os_type']
+    windows = os_type == 'windows'
+
+    error_context.context('Boot guest %s on src host.' % src_desc, logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=360)
+    if windows:
+        session = utils_test.qemu.windrv_check_running_verifier(
+            session, vm, test, params["driver_name"])
+
+    if with_cdrom:
+        cdrom_params = params.object_params(params['cdroms'])
+        check_orig_items = ast.literal_eval(cdrom_params['check_orig_items'])
+        dev_id = check_orig_items['device']
+        check_cdrom_info_by_qmp(check_orig_items)
+        orig_size = compare_cdrom_size(params['cdrom_orig_file'])
+
+        orig_img_name = params["cdrom_orig_file"]
+        device_name = vm.get_block({"file": orig_img_name})
+        if device_name is None:
+            test.fail("Failed to get device using image %s." % orig_img_name)
+
+        eject_check = QMPEventCheckCDEject(vm, device_name)
+        change_check = QMPEventCheckCDChange(vm, device_name)
+        eject_cdrom()
+        change_cdrom()
+
+        check_new_items = ast.literal_eval(cdrom_params['check_new_items'])
+        check_cdrom_info_by_qmp(check_new_items)
+        new_size = compare_cdrom_size(params['cdrom_new_file'])
+        if new_size == orig_size:
+            test.fail('The new size inside guest is equal to the orig iso size.')
+
```

Comment:
session is not used after this, maybe need a session.close() ?

Reply:
Don't worry about it, the framework will close all the sessions in postprocess_vm, so you wouldn't need to close it in case.

```
def postprocess_vm(test, params, env, name):
    """
    Postprocess a single VM object according to the instructions in params.
    Kill the VM if requested and get a screendump.

    :param test: An Autotest test object.
    :param params: A dict containing VM postprocessing parameters.
    :param env: The environment (a dict-like object).
    :param name: The name of the VM object.
    """
    vm = env.get_vm(name)
    if not vm:
        return

    # Close all SSH sessions that might be active to this VM
    for s in vm.remote_sessions[:]:
        try:
            s.close()
            vm.remote_sessions.remove(s)
        except Exception:
            pass
```